### ISSUE TITLE
no-extra-semicolons

### ISSUE THREAD

**Issue by justo** [stylelint/stylelint#1103]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A new rule to disallow extra trailing semicolons. Essentially http://eslint.org/docs/rules/no-extra-semi

> Which rule, if any, is this issue related to?

None at the moment.

> What CSS is needed to reproduce this issue?

The following will _not_ be an error

```css
.foo { color: pink;;;;;; }

.bar {
  color: pink;
  ;;
}
```

Although other errors may be triggered by having extra trailing semicolons:
```css
.foo {
  color: pink;;;
}
```
Will throw `block-closing-brace-newline-before` if enabled.

> What stylelint configuration is needed to reproduce this issue?

`stylelint-config-standard@6.0.0`

```json
{
	"extends": "stylelint-config-standard"
}
```

> Which version of stylelint are you using?

6.0.3

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI `stylelint  "./src/**/*.scss" --syntax "scss"`
and Atom plugin `linter-stylelint`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

This happens when testing SCSS and normal CSS.

> What did you expect to happen?

Should probably be an error with the standard config.

> What actually happened (e.g. what warnings or errors you are getting)?

Nothing, `.foo { color: pink;;;; }` passes.

-----
I'd be happy to build the rule if you think it should be implemented.

**Comment by jeddy3** [comment]
@justo Thanks for the suggestion, it's a good one!

I'm wondering whether we should broaden this out to checking for extra semicolons *anywhere in the sheetsheet*, rather than *just within declaration blocks*. Mainly because I think it'll be tricker to build companion rules specific to, say, extra semicolons after an at-rule statement (e.g. `@import "x.css";;;`) as [PostCSS stores the extra semicolons in various `raw.after/before` properties](https://astexplorer.net/#/x2qqRpFclL/2).

If we were to go with something like `no-extra-semicolons`, then I think it'd be the case of looking through the `raw.after/before` properties of *every* node (including the `root` one) for the occurrence of a semicolon. We wouldn't then need to distinguish between statements, comments and declarations using this approach.

Anyone see any issues with this idea?

> I'd be happy to build the rule if you think it should be implemented.

That's great!  There's a [guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) to help you get started. If we do go down the `no-extra-semicolons` route, then the most similar existing rule is probably [`no-invalid-double-slash-comments`](https://github.com/stylelint/stylelint/tree/master/src/rules/no-invalid-double-slash-comments) and you can get ideas for options validation, docs, messages and tests from there. 

It's best to wait until people have had time to feedback on the inclusion of this rule before making a start though :)

**Comment by davidtheclark** [comment]
> If we were to go with something like no-extra-semicolons, then I think it'd be the case of looking through the raw.after/before properties of every node (including the root one) for the occurrence of a semicolon.

Actually, I think this might then be like `string-quotes` where you're not actually searching through nodes but just scanning the whole text, looking for occurrences of multiple semicolons in a row. This can be done with the `styleSearch()` utility.

**Comment by davidtheclark** [comment]
Forgot to say: Yeah, I'm in favor a `no-extra-semicolons` rule!

**Comment by jeddy3** [comment]
> looking for occurrences of multiple semicolons in a row.

Should the rule also check for single misplaced semicolons? e.g.

```css
a {}
;
b {}
```

PostCSS [seems to parse that](https://astexplorer.net/#/5uq2rJF6D9).

Would this be possible with `styleSearch()`?

**Comment by davidtheclark** [comment]
`styleSearch()` doesn't know what node the semicolon is in. So not sure how that should be targeted. I'd say whichever method makes the most tests pass is the one we should use!

**Comment by jeddy3** [comment]
> I'd say whichever method makes the most tests pass is the one we should use!

Sounds good to me :)

@justo Feel free make a start on this rule if you like. It isn't immediately clear which implementation would work best: checking the `raw` properties or using the `styleSearch` utility. So, you might want to experiment a little. You might also want to consider running [performance benchmarks](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/benchmarks.md) as you experiment in case there are any significant performance differences between the implementations.

In terms of tests, here are a few ideas for what you might be checking for:

```css
;
@import "x.css";;
@import "y.css"; ;

.foo { color: pink;;;;;; }
;
.bar {
  color: pink;
  ;;
  ;
}

;;;

/* comments */

;;;
```

Feel free to ask questions here, or open a WIP pull requests if you have any queries :)

**Comment by evilebottnawi** [comment]
@jeddy3 how about spaces?
```css
a {
   color: red; ;
}
```

**Comment by jeddy3** [comment]
>@jeddy3 how about spaces?

Already included on Line 3 of the example CSS :)

**Comment by evilebottnawi** [comment]
@jeddy3 oh sorry

**Comment by justo** [comment]
Thanks for the direction and tips on how to get started. Glad you support adding the rule. I'm going to start poking around and see what I can come up with. I'm sure I'll have some questions for you guys.

**Comment by evilebottnawi** [comment]
Merge. Close.

**Comment by justo** [comment]
Darn I had done a fair bit of work on this already. Oh well, at least the rule is in now.

**Comment by evilebottnawi** [comment]
@justo oh sorry, i need this rule too :cry:, write here if you start work, we change label on `wip` and all know about this rule will create
